zbus_xmlgen/src/main.rs

```diff
@@ -19,112 +18,162 @@ use zbus_xml::{Interface, Node};
-        Some(help) if help == "--help" || help == "-h" => {
-            usage();
-            return Ok(());
-        }
-        Some(path) => {
-            input_src = Path::new(&path)
-                .file_name()
-                .unwrap()
-                .to_string_lossy()
-                .to_string();
+    let (node, service, path, input_src) = match args.command {
+        cli::Command::System {
+            service,
+            object_path,
+        } => build_dbus_info(Connection::system()?, service, object_path)?,
+        cli::Command::Session {
+            service,
+            object_path,
+        } => build_dbus_info(Connection::session()?, service, object_path)?,
+        cli::Command::Address {
+            address,
+            service,
+            object_path,
+        } => build_dbus_info(
+            connection::Builder::address(&*address)?.build()?,
+            service,
+            object_path,
+        )?,
+        cli::Command::File { path } => {
+            let input_src = path.file_name().unwrap().to_string_lossy().to_string();
             let f = File::open(path)?;
-            (Node::from_reader(f)?, None, None)
-        }
-        None => {
-            usage();
-            return Ok(());
+            (Node::from_reader(f)?, None, None, input_src)
         }
     };
 
-    let mut process = match Command::new("rustfmt").stdin(Stdio::piped()).spawn() {
-        Err(why) => panic!("couldn't spawn rustfmt: {}", why),
-        Ok(process) => process,
-    };
-    let rustfmt_stdin = process.stdin.as_mut().unwrap();
     let fdo_iface_prefix = "org.freedesktop.DBus";
     let (fdo_standard_ifaces, needed_ifaces): (Vec<&Interface<'_>>, Vec<&Interface<'_>>) = node
         .interfaces()
         .iter()
         .partition(|&i| i.name().starts_with(fdo_iface_prefix));
 
-    if let Some((first_iface, following_ifaces)) = needed_ifaces.split_first() {
+    if !args.split_interfaces {
+        let output = write_interfaces(
+            &needed_ifaces,
+            &fdo_standard_ifaces,
+            &service,
+            &path,
+            &input_src,
+        )?;
+        println!("{}", output);
+        return Ok(());
+    }
+
+    for interface in needed_ifaces {
+        let output = write_interfaces(
+            &[&interface],
+            &fdo_standard_ifaces,
+            &service,
+            &path,
+            &input_src,
+        )?;
+        std::fs::write(
+            format!(
+                "{}.rs",
+                interface
+                    .name()
+                    .split('.')
+                    .last()
+                    .expect("Failed to split name")
+            ),
+            output,
+        )?;
+    }
+
+    Ok(())
+}
+
+fn build_dbus_info<'a>(
+    connection: Connection,
+    service: String,
+    object_path: String,
+) -> Result<
+    (
+        Node<'a>,
+        Option<BusName<'a>>,
+        Option<ObjectPath<'a>>,
+        String,
+    ),
+    Box<dyn Error>,
```

Comment:
This results in a [`type_complexity` clippy warning](https://rust-lang.github.io/rust-clippy/master/index.html#/type_complexity). Not sure if it makes sense to create a type just for this, since the returned tuple is immediately destructured.

Reply:
creating a type is so trivial that I always heed to this warning. :) Coming up with names is the issue. :)